fix: don't create aws_kms keys when in check mode

Tyler-2 opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Fixes https://github.com/ansible/ansible/issues/68019

https://github.com/ansible/ansible/issues/68019
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
aws_kms

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
When running aws_kms for key creation in check mode, check mode is ignored and the key is actually created. This commit solves this. See issue description for more details.
<!--- Paste verbatim command output below, e.g. before and after your change -->
```
(venv) tyler@func-platform:~/Repos/ansible-fork$ ansible-playbook tyler-test.yml --check
TASK [Create KMS key] *********************************************************************************************************************************************************************************************
task path: /home/tyler/Repos/ansible-fork/tyler-test.yml:5
changed: [localhost] => {
    "changed": true, 
    "invocation": {
        "module_args": {
            "alias": "test-kms-key", 
            "aws_access_key": null, 
            "aws_config": null, 
            "aws_secret_key": null, 
            "debug_botocore_endpoint_logs": false, 
            "description": null, 
            "ec2_url": null, 
            "enable_key_rotation": null, 
            "enabled": true, 
            "grants": [], 
            "key_id": null, 
            "policy": null, 
            "policy_clean_invalid_entries": true, 
            "policy_grant_types": null, 
            "policy_mode": "grant", 
            "policy_role_arn": null, 
            "policy_role_name": null, 
            "profile": null, 
            "purge_grants": false, 
            "purge_tags": false, 
            "region": null, 
            "security_token": null, 
            "state": "present", 
            "tags": {}, 
            "validate_certs": true
        }
    }
}
META: ran handlers
META: ran handlers

PLAY RECAP ********************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

(venv) tyler@func-platform:~/Repos/ansible-fork$ aws kms list-aliases | grep test-kms-key
(venv) tyler@func-platform:~/Repos/ansible-fork$ 
```

Tyler-2 commented:
@tremble I think I have added sane ftests (though I'm not sure where they run, since it doesn't look like shippable knows about them). Also made the change to avoid the big conditional block, since that was trivial.

Tyler-2 commented:
Checking into ftest failure

flowerysong commented on plugins/modules/aws_kms.py:
```diff
@@ -821,18 +821,19 @@ def create_key(connection, module):
     if module.params.get('policy'):
         params['Policy'] = module.params['policy']
 
+    if module.check_mode:
```
This should go at the very top of the function.

Tyler-2 replied:
I kind of like to run as much of the module as possible, up until the thing that actually does the change. Is this unconventional with check mode?

tremble replied:
depends on the author, I tend to keep it as 'late' as possible too.

flowerysong replied:
In this particular case those few lines don't actually do anything, so I prefer having it at the beginning to make it clear that the function doesn't do anything.

Tyler-2 replied:
Made this change.

flowerysong commented on plugins/modules/aws_kms.py:
```diff
@@ -821,18 +821,19 @@ def create_key(connection, module):
     except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
         module.fail_json_aws(e, msg="Failed to create initial key")
     key = get_key_details(connection, module, result['KeyId'])
-
```
I don't think this blank line should be removed.

Tyler-2 replied:
Inadvertent when I made the restruct change. Corrected.

flowerysong commented on plugins/modules/aws_kms.py:
```diff
@@ -821,18 +821,19 @@ def create_key(connection, module):
 
     ensure_enabled_disabled(connection, module, key, module.params.get('enabled'))
     update_grants(connection, module, key, module.params.get('grants'), False)
-
```
Same here.

Tyler-2 replied:
Fixed.

flowerysong commented on tests/integration/targets/aws_kms/tasks/main.yml:
```diff
@@ -32,6 +32,26 @@
+          Hello: World
+        state: present
+        enabled: yes
+
```
Needs an `assert` task to verify that the module correctly returned 'changed'.

Tyler-2 replied:
Addressed.

flowerysong commented on tests/integration/targets/aws_kms/tasks/main.yml:
```diff
@@ -65,6 +85,13 @@
+      check_mode: yes
+      aws_kms:
+        alias: "{{ resource_prefix }}-kms"
+        state: absent
```
Also needs an `assert`.

Tyler-2 replied:
Addressed.

Tyler-2 commented:
@flowerysong I think I addressed most of your concerns, but I generally let the requestor hit the "resolve conversations". Let me know if there's something more.

flowerysong commented:
@Tyler-2 People don't normally have to ability to resolve those for you, unless they have extra privileges on the repository itself.

I don't.
